src/Kiota.Builder/Writers/Shell/ShellCodeMethodWriter.cs

```diff
@@ -0,0 +1,405 @@
+        private const string outputFormatterFactoryParamName = "outputFormatterFactory";
+
+        public ShellCodeMethodWriter(CSharpConventionService conventionService) : base(conventionService)
+        {
+        }
+
+        protected override void WriteCommandBuilderBody(CodeMethod codeElement, RequestParams requestParams, bool isVoid, string returnType, LanguageWriter writer)
+        {
+            var parent = codeElement.Parent as CodeClass;
+            var classMethods = parent.Methods;
+            var name = codeElement.SimpleName;
+            name = uppercaseRegex.Replace(name, "-$1").TrimStart('-').ToLower();
+
+            if (codeElement.HttpMethod == null)
+            {
+                // Build method
+                // Puts together the BuildXXCommand objects. Needs a nav property name e.g. users
+                // Command("users") -> Command("get")
+                if (string.IsNullOrWhiteSpace(name))
+                {
+                    // BuildCommand function
+                    WriteUnnamedBuildCommand(codeElement, writer, parent, classMethods);
+                }
+                else
+                {
+                    WriteContainerCommand(codeElement, writer, parent, name);
+                }
+            } else
+            {
+                WriteExecutableCommand(codeElement, requestParams, writer, name);
+            }
+        }
+
+        private void WriteExecutableCommand(CodeMethod codeElement, RequestParams requestParams, LanguageWriter writer, string name)
+        {
+            var generatorMethod = (codeElement.Parent as CodeClass)
+                                           .Methods
+                                           .FirstOrDefault(x => x.IsOfKind(CodeMethodKind.RequestGenerator) && x.HttpMethod == codeElement.HttpMethod);
+            var pathAndQueryParams = generatorMethod.PathAndQueryParameters;
+            var originalMethod = codeElement.OriginalMethod;
+            var origParams = originalMethod.Parameters;
+            var parametersList = pathAndQueryParams?.Where(p => !string.IsNullOrWhiteSpace(p.Name))?.ToList() ?? new List<CodeParameter>();
+            if (origParams.Any(p => p.IsOfKind(CodeParameterKind.RequestBody)))
+            {
+                parametersList.Add(origParams.OfKind(CodeParameterKind.RequestBody));
+            }
+            writer.WriteLine($"var command = new Command(\"{name}\");");
+            WriteCommandDescription(codeElement, writer);
+            writer.WriteLine("// Create options for all the parameters");
+            // investigate exploding query params
+            // Check the possible formatting options for headers in a cli.
+            // -h A=b -h
+            // -h A:B,B:C
+            // -h {"A": "B"}
+            var availableOptions = new List<string>();
+            foreach (var option in parametersList)
+            {
+                var type = option.Type as CodeType;
```

Comment:
here we have a good opportunity to move this block to a separate method and reduce cognitive complexity